py/server/deephaven/exec_ctx.py

```diff
@@ -0,0 +1,87 @@
+
+
+class ExecutionContext(JObjectWrapper):
+    """An ExecutionContext represents a specific combination of query library, query compiler, and query scope under
+    which a query is evaluated. An implicit, systemic, default ExecutionContext is created for all immediate query
+    evaluations. However, to be able to evaluate a query in a deferred manner, such as in a separate user created or
+    engine created thread, an application must explicitly install a ExecutionContext on the thread or enclose the query
+    inside the body of the with statement on a ExecutionContext.
+    """
+    j_object_type = _JExecutionContext
+    default_ctx: _JExecutionContext
+
+    @property
+    def j_object(self) -> jpy.JType:
+        return self.j_exec_ctx
+
+    def __init__(self, j_exec_ctx):
+        self.j_exec_ctx = j_exec_ctx
+
+    def __enter__(self):
+        self._j_safe_closable = self.j_exec_ctx.open()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._j_safe_closable.close()
+
+    @staticmethod
+    def get_exec_ctx() -> ExecutionContext:
+        """Returns the current thread's ExecutionContext.
+
+        Returns:
+            a ExecutionContext
+        """
+        return ExecutionContext(j_exec_ctx=_JExecutionContext.getContext())
+
+    @staticmethod
+    def set_exec_ctx(exec_ctx: ExecutionContext) -> None:
+        """Sets the current thread's ExecutionContext.
+
+        Args:
+            exec_ctx (ExecutionContext): the ExecutionContext to be installed on the current thread
+        """
+        _JExecutionContext.setContext(exec_ctx.j_exec_ctx)
+
+    @classmethod
+    def _clone_ctx(cls) -> ExecutionContext:
+        """Returns a cloned ExecutionContext of the current thread's ExecutionContext.
+
+         Question: the imagined main use case is to clone the implicit good EC for use in a thread, but if the
+         implicit good one is thread-safe, then why bother?
+        """
+        j_exec_ctx = (_JExecutionContext.newBuilder()
+                      .captureQueryCompiler()
+                      .captureQueryLibrary()
+                      .captureQueryScope()
+                      .build())
+        return ExecutionContext(j_exec_ctx=j_exec_ctx)
+
+    @classmethod
+    def _build_ctx(cls, j_query_compiler, j_query_library, j_query_scope) -> ExecutionContext:
+        """Builds a ExecutionContext
+
+        Question: without wrapping these raw Java types and without useful alternative options in the server for
+        these types, this method doesn't make much sense. leave out for now?
+        """
+        j_exec_ctx = (_JExecutionContext.newBuilder()
+                      .setQueryCompiler(j_query_compiler)
+                      .setQueryLibrary(j_query_library)
+                      .setQueryScope(j_query_scope)
+                      .build())
+        return ExecutionContext(j_exec_ctx=j_exec_ctx)
+
```

Comment:
@nbauernfeind  Please let me know if my thinking in the two questions makes sense?